### PR TITLE
[SYCLomatic] Refine the logic to remove redundant wait after memcpy

### PR DESCRIPTION
**Pull request by intwanghao**

### CODE REVIEW

**Review by wenhuiNi — APPROVED** [review]

**Review by zhiweij1 — APPROVED** [review]

File: clang/test/dpct/memcpy_optimization.cu [hunk 1]
```diff
@@ -13,5 +16,44 @@ int main(){
     cudaMemcpy(a, b, 10, cudaMemcpyHostToDevice);
     cudaMemcpy(a, b, 10, cudaMemcpyDeviceToDevice);
 
+//CHECK:  for(int i = 0; i < 10; i++) {
+//CHECK:    int src;
+//CHECK:    q_ct1.memcpy(a, &src, 10).wait();
+//CHECK:  }
+    for(int i = 0; i < 10; i++) {
+      int src;
```
Comment (zhimingwang36) [inline clang/test/dpct/memcpy_optimization.cu:24]:
```suggestion
      int src=i; // update src content for each iter
```

File: clang/test/dpct/memcpy_optimization.cu [hunk 2]
```diff
@@ -13,5 +16,44 @@ int main(){
...
+      cudaMemcpy(a, &src, 10, cudaMemcpyHostToDevice);
```
Comment (zhimingwang36) [inline clang/test/dpct/memcpy_optimization.cu:25]:
```suggestion
      cudaMemcpy(dev_a, &src, 10, cudaMemcpyHostToDevice);
```

File: clang/test/dpct/memcpy_optimization.cu [hunk 3]
```diff
@@ -13,5 +16,44 @@ int main(){
...
+//CHECK:  }
+    if(cudaMemcpy(a, e, 10, cudaMemcpyHostToDevice)) {
+        std::cout << "failed" << std::endl;
+    }
+
+//CHECK:  q_ct1.memcpy(a, f, 10).wait();
+//CHECK:  if(sync) {
+//CHECK:    dev_ct1.queues_wait_and_throw();
+//CHECK:  }
+    free(f);
```
Comment (zhimingwang36) [inline clang/test/dpct/memcpy_optimization.cu:51]:
```suggestion

```

File: clang/test/dpct/USM-restricted.cu [hunk 4]
```diff
@@ -132,20 +132,20 @@ void foo() {
 
   // CHECK: q_ct1.memcpy(d_A, h_A, size).wait();
   cudaMemcpy(d_A, h_A, size, cudaMemcpyHostToDevice);
```
Comment (zhimingwang36) [inline clang/test/dpct/USM-restricted.cu:134]: can be further refine, Not sure this will hurt perf.

**Pull request closed without merging**